### PR TITLE
Add TypeScript support

### PR DESCRIPTION
**Pull request by basarat**
🌹

### CODE REVIEW

**Comment by ai** [comment]
Great!

Can you also send types for `generate` and `format` functions?

**Comment by basarat** [comment]
Done :rose:

**Comment by ai** [comment]
@basarat Looks great! Few question:

1. What is `.alm` in `.gitignore`?
2. Do we need `tsconfig.json`?
3. Do we need to mention other `.d.ts` files in `package.json`?

**Comment by basarat** [comment]
> What is .alm in .gitignore?

Just what alm uses http://alm.tools/ like `.idea` or `.vscode` etc. 

> Do we need tsconfig.json?

Helps make sure there isn't a TypeScript error in the `.d.ts` files 

> Do we need to mention other .d.ts files in package.json?

No. As `import * as generate from "nonoid/generate"` will automatically lookup a `generate.d.ts` file in the `node_modules/nanoid` folder :rose:

**Comment by ai** [comment]
Great. Can you remove `.alm`?

If we have `tsconfig`, should we use it in `npm test`? Should we also add this config to `.npmignore`?

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+declare const nanoid: () => string;
```
Comment (BrainMaestro) [inline index.d.ts]: `nanoid` takes a [number parameter](https://github.com/ai/nanoid/blob/ef1d664b52fb0cd978371c9601658f09c84c88cc/index.js#L21-L22)

File: generate.d.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+declare const generate: (charset: string, length: number) => string;
```
Comment (BrainMaestro) [inline generate.d.ts]: the actual [`generate`](https://github.com/ai/nanoid/blob/ef1d664b52fb0cd978371c9601658f09c84c88cc/generate.js#L21) functions's arguments are `alphabet` and `size`.

File: format.d.ts [hunk 3]
```diff
@@ -0,0 +1,2 @@
+declare const format: (random: (size: number) => string[], alphabet: string, size: number) => string;
+export = format;
```
Comment (TrySound) [inline format.d.ts]: Why not `export default`?

**Comment by TrySound** [comment]
@basarat Will typescript find `index.d.ts` with `import nanoid from 'nanoid';`?

File: format.d.ts (hunk 3 above)
Comment (TrySound) [inline format.d.ts]: All modules here has only one export so you may don't use tricky funky synthetic exports.

**Comment by ai** [comment]
Typings repositories sounds like good idea :) In this case only Flow/TS users will download type definitions
https://github.com/ai/nanoid/pull/15

**Pull request closed without merging**